paasta_tools/mesos_maintenance.py

```diff
@@ -331,7 +399,110 @@ def load_credentials(mesos_secrets='/nail/etc/mesos-slave-secret'):
+    :param slave_id: the id of the mesos slave
+    :param resources: list of Resource named tuples specifying the name and amount of the resource to (un)reserve
+    :returns: boolean where 0 represents success and 1 is a failure
+    """
+    log.info("Dynamically reserving resoures on %s: %s" % (slave_id, resources))
+    payload = {
+        'slaveId': slave_id,
+        'resources': str(build_reservation_payload(resources)).replace("'", '"').replace('+', '%20')
+    }
+    client_fn = reserve_api()
+    try:
+        reserve_output = client_fn(method="POST", endpoint="", data=payload).text
+    except HTTPError:
+        log.debug("Error adding dynamic reservation.")
+        raise
+    return reserve_output
+
+
+def unreserve(slave_id, resources):
+    """Dynamically unreserve resources in marathon to allow tasks to using them.
+    :param slave_id: the id of the mesos slave
+    :param resources: list of Resource named tuples specifying the name and amount of the resource to (un)reserve
+    :returns: boolean where 0 represents success and 1 is a failure
+    """
+    log.info("Dynamically unreserving resoures on %s: %s" % (slave_id, resources))
+    payload = {
+        'slaveId': slave_id,
+        'resources': str(build_reservation_payload(resources)).replace("'", '"').replace('+', '%20')
+    }
+    client_fn = unreserve_api()
+    try:
+        unreserve_output = client_fn(method="POST", endpoint="", data=payload).text
+    except HTTPError:
+        log.debug("Error adding dynamic unreservation.")
+        raise
+    return unreserve_output
+
+
+def reserve_all_resources(hostnames):
+    mesos_state = get_mesos_master().state_summary()
+    hosts = []
+    components = hostnames_to_components(hostnames)
+    for component in components:
+        hosts.append(component.host)
+    for slave in mesos_state['slaves']:
+        hostname = slave['hostname']
+        if hostname in hosts:
+            log.info("Reserving all resources on %s" % hostname)
+            slave_id = slave['id']
+            resources = []
+            for resource in ['disk', 'mem', 'cpus']:
+                free_resource = slave['resources'][resource] - slave['used_resources'][resource]
+                for role in slave['reserved_resources']:
+                    free_resource -= slave['reserved_resources'][role][resource]
+                resources.append(Resource(name=resource, amount=free_resource))
+            try:
+                reserve(slave_id=slave_id, resources=resources)
+            except HTTPError:
+                log.debug("Failed reserving all of the resources on %s (%s). Aborting." % (hostname, slave_id))
+                raise
+
+
+def unreserve_all_resources(hostnames):
+    mesos_state = get_mesos_master().state_summary()
+    hosts = []
+    components = hostnames_to_components(hostnames)
+    for component in components:
+        hosts.append(component.host)
```

Comment:
the logic here being duplicated from ``reserve_all_resources`` hints that you should move it into its own function with a unit test